### PR TITLE
Store binding credentials in secretsmanager not cloudformation outputs

### PR DESCRIPTION
**Pull request by chrisfarms**
## What

* stops writing access key/secret values to cloudformation outputs
* writes a JSON blob for the binding credentials into ~~parameter store~~ secretsmanager as part of the binding stack
* `GetBinding` now no-longer requires fetching both stacks and only needs access to the binding stack to discover the ~~paramstore path~~ secretsmanager arn of the credentials
* credential keys are now in `snake_case_form` for consistency with the rest of PaaS brokers
* credential keys now match those of s3 broker's IAM creds for consistancy
* renamed `main` -> `primary` ... and `dl` -> `secondary` .... this makes it clearer throughout the code/tests (all the "DL" was hard to spot after coming back to it - and primary vs secondary should be more intuitive for getting started)
* now pulls in ~~SSM~~ secretsmanager client, fakes updated, etc

## Note

* Since the credential names/format have changed, the platform-tests in paas-cf will need updating.

### CODE REVIEW

File: sqs/provider.go [hunk 1]
```diff
@@ -364,37 +341,19 @@ func (s *Provider) GetBinding(ctx context.Context, getBindingData provideriface.
...
-	} else if err != nil {
-		// failed to get stack status
+	credentialsARN := getStackOutput(userStack, OutputCredentialsARN)
+	res, err := s.Client.GetSecretValueWithContext(ctx, &secretsmanager.GetSecretValueInput{
+		SecretId: aws.String(credentialsARN),
+	})
+	if err != nil {
 		return nil, err
+	} else if res.SecretString == nil {
+		return nil, fmt.Errorf("invalid response from parameter store")
```
Comment (Krenair) [inline sqs/provider.go:351]:
```suggestion
		return nil, fmt.Errorf("invalid response from secrets manager")
```

**Comment by Krenair** [comment]
otherwise looks good to me

File: sqs/provider.go (hunk 1 above)
Comment (chrisfarms) [reply sqs/provider.go]: fixedup

**Comment by risicle** [comment]
I assume this has already been discussed, but the only hesitation I'd have about the "dead letter" -> "secondary" switch is if you google "sqs secondary queue", you basically get zero useful results. Google "sqs dead letter queue" it tells you all about it. Inventing a new terminology could sort of divorce this from the mainstream documentation.

**Comment by chrisfarms** [comment]
> I assume this has already been discussed, but the only hesitation I'd have about the "dead letter" -> "secondary" switch is if you google "sqs secondary queue", you basically get zero useful results. Google "sqs dead letter queue" it tells you all about it. Inventing a new terminology could sort of divorce this from the mainstream documentation.

Thing is. It's _not_ a dead-letter queue.... Until it is.

We provision you two queues... They are just two identical queues.

Then, _optionally_ you can perform an update to the service and you can configure messages that fail to get consumed on the primary queue to end up on the secondary queue (thus turning it into a dead-letter queue.

We will explicitly document this use case in the paas-docs, so users who require this functionality will easily be able to set it up as a dead-letter queue with a suitable restive policy, so I'm not concerned with people not immediately knowing what the secondary queue is for.... I'm actually _more_ concerned with having something named "dead-letter-queue" that actually _isnt_ a dead letter queue!

**Comment by Krenair** [comment]
I was sceptical at first but yeah, agreed

**Review by Krenair — APPROVED** [review]

**Comment by risicle** [comment]
Ok, well that's an interesting quirk :+1:

So if I understand this correctly:

There's no point in _not_ creating it at provision time if it's not needed because unused SQS queues are effectively free and it reduces the complexity of the actions you have to take on an update event.

But setting up a functioning dead letter queue for everyone whether they need it or not isn't free.

And you're not omitting the `SecondaryQueue*` keys when DL isn't set up... because it's actually painful to have to go and figure out whether it has DL set up or not every time you're asked to return service details?

**Comment by chrisfarms** [comment]
> Ok, well that's an interesting quirk

Yeah... it's definitly a bit weird having two queues, but the alternative would be to allow somehow wiring separate cloudfoundry services to each other... which would be a really horrible UX, involving extracting the queue ARNs from service status and passing it back in when creating another queue.

Basically this "two queues" thing is a way to make the UX for the very common case of "I want unprocessable items to end up on a different queue" a bit nicer via the cloudfoundry interface we are stuck behind.

> There's no point in _not_ creating it at provision time if it's not needed because unused SQS queues are effectively free and it reduces the complexity of the actions you have to take on an update event.
>
> But setting up a functioning dead letter queue for everyone whether they need it or not isn't free.

The queues are effectivitly free yes, however it's a little more complex than just "setting up a dead letter queue isn't free"...

There isn't a "one size fits all" for dead-letter queues, the settings will be very specific to your use case. For example you may want ALL items that fail to process once to get pulled off the primary queue, OR you may want to retry things N times first and then later give up and put them on a different queue, OR other more complex setups.... so it isn't practical to say "here we set you up a dead-letter queue" without making large assumptions about it's use. This means configuration must be optional and left to the user.

In addition to that, you also don't want the act of enabling/disabling the configuration (redrive policy) of sending items to a dead-letter queue to cause the creation/destruction of the secondary queue, because if you later decide to change your settings, you may end up destroying the secondary queue with items still on it! So you want to have your queues ready from the start and not have to worry about them getting destroyed when you change your redrive policy.

> And you're not omitting the `SecondaryQueue*` keys when DL isn't set up... because it's actually painful to have to go and figure out whether it has DL set up or not every time you're asked to return service details?

There are not separate credentials for the queues... the same access key/secret give you access to both queues.

If you aren't using the secondary queue as a dead-letter queue, you can use it for something else if you like using the same creds.

**Comment by risicle** [comment]
Ok thanks this is making sense... It's just dawning on me now that the biggest thing cloudformation is doing for us in this broker is giving us a way of addressing all of our created resources under the _one_ "instance id" that cloudfoundry will supply us with in requests.

**Pull request merged**